### PR TITLE
Add audit filter for write-request body logging

### PR DESCRIPTION
**Pull request by gsleeman**
This is for getting the audit-log-exporter binary into the splunk-forwarder image

### CODE REVIEW

File: audit-filter/README.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+* `openshift-*` service account write events that occur within the same namespace are __dropped__
+* Requests that had a response code of 409, 422 or 404 are __dropped__
+* All other events are __forwarded__
+
+## Deduplication
+
+The process by which `update` events that include the full object in the `requestBody` are deduplicated is:
+
+1. The `requestObject.metadata.resourceVersion` and `requestObject.metadata.generation` fields are removed (resourceVersion remains present at `objectRef.resourceVersion`)
+2. The event is then to a `patch` request by generating a three-way JSON patch based on the previous cached `update` request, if one exists.
```
Comment (2uasimojo) [inline audit-filter/README.md:37]: Missing verb here, I think.

```suggestion
2. The event is then converted to a `patch` request by generating a three-way JSON patch based on the previous cached `update` request, if one exists.
```
?

File: audit-filter/README.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# audit-log-exporter
+Enables policy-driven event filtering and deduplication of OpenShift audit logs for splunk-forwarder
+
+## Enabling in stage
+
+Forwarding audit logs to `openshift_managed_audit_stage` can be enabled by adding a policy file to the `osd-monitored-logs-local` configmap:
+
+```shell
+oc set data cm/osd-monitored-logs-local --from-file=policy.yaml
+```
```
Comment (2uasimojo) [inline audit-filter/README.md:10]: Unclear whether the [supplied policy.yaml file](https://github.com/openshift/boilerplate/blob/master/README.md#update) is intended for this purpose; and if it is, whether it must/should be edited first.

File: audit-filter/README.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+## Deduplication
+
+The process by which `update` events that include the full object in the `requestBody` are deduplicated is:
+
+1. The `requestObject.metadata.resourceVersion` and `requestObject.metadata.generation` fields are removed (resourceVersion remains present at `objectRef.resourceVersion`)
+2. The event is then to a `patch` request by generating a three-way JSON patch based on the previous cached `update` request, if one exists.
+3. The event is discarded a patch was generated but itsbody is empty
+
+Events converted from `update` to `patch` (but _not_ discarded) are annotated indicating the conversation. Metrics for discarded events are recorded in `events_processed_total{decision="drop",reason="no-op write"}`
```
Comment (2uasimojo) [inline audit-filter/README.md:40]:
```suggestion
Events converted from `update` to `patch` (but _not_ discarded) are annotated indicating the conversion. Metrics for discarded events are recorded in `events_processed_total{decision="drop",reason="no-op write"}`
```
?

File: containers/forwarder/Dockerfile [hunk 4]
```diff
@@ -1,3 +1,8 @@
+FROM registry.ci.openshift.org/openshift/release:golang-1.13 AS builder
+COPY audit-filter /go/src/audit-filter
+WORKDIR /go/src/audit-filter
+RUN make
```
Comment (2uasimojo) [inline containers/forwarder/Dockerfile:4]: Suggest being explicit about the target, so Makefile edits don't have unintended repercussions.
```suggestion
RUN make build
```

File: audit-filter/README.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
+# audit-log-exporter
```
Comment (2uasimojo) [inline audit-filter/README.md:1]: Should link this from the root README.

**Review by 2uasimojo — COMMENTED** [review]
NB: I didn't review the actual code. Because what I'm here for is...

We need to change some things about the building and tagging system if PRs like this are going to work. The reason is that images are [currently](https://github.com/openshift/splunk-forwarder-images/blob/296e8fd02851a03f0189e693d9cf23602f3364e1/Makefile#L16) [tagged](https://github.com/openshift/splunk-forwarder-images/blob/296e8fd02851a03f0189e693d9cf23602f3364e1/Makefile#L20) based *only* on the splunk universal forwarder version; and [if that version doesn't change, we don't rebuild/repush the splunk-(heavy)forwarder images](https://github.com/openshift/splunk-forwarder-images/blob/296e8fd02851a03f0189e693d9cf23602f3364e1/hack/app-sre-build-push.sh#L83-L84) at all.

I suggest is something like appending the current hash of this repo, like we do for operator versions. The delta would be [here](https://github.com/openshift/splunk-forwarder-images/blob/296e8fd02851a03f0189e693d9cf23602f3364e1/Makefile#L12-L21) and might be:

```diff
diff --git a/Makefile b/Makefile
index 7e9d719..8c608ba 100644
--- a/Makefile
+++ b/Makefile
@@ -11,13 +11,15 @@ CONTAINER_ENGINE=$(shell command -v podman 2>/dev/null || command -v docker 2>/d
 
 FORWARDER_VERSION=$(shell cat .splunk-version)
 FORWARDER_HASH=$(shell cat .splunk-version-hash)
+CURRENT_COMMIT=$(shell git rev-parse --short=7 HEAD)
+IMAGE_TAG=$(FORWARDER_VERSION)-$(FORWARDER_HASH)-$(CURRENT_COMMIT)
 
 FORWARDER_NAME=splunk-forwarder
-FORWARDER_IMAGE_URI=$(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$(FORWARDER_NAME):$(FORWARDER_VERSION)-$(FORWARDER_HASH)
+FORWARDER_IMAGE_URI=$(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$(FORWARDER_NAME):$(IMAGE_TAG)
 FORWARDER_DOCKERFILE = ./containers/forwarder/Dockerfile
 
 HEAVYFORWARDER_NAME=splunk-heavyforwarder
-HEAVYFORWARDER_IMAGE_URI=$(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$(HEAVYFORWARDER_NAME):$(FORWARDER_VERSION)-$(FORWARDER_HASH)
+HEAVYFORWARDER_IMAGE_URI=$(IMAGE_REGISTRY)/$(IMAGE_REPOSITORY)/$(HEAVYFORWARDER_NAME):$(IMAGE_TAG)
 HEAVYFORWARDER_DOCKERFILE = ./containers/heavy_forwarder/Dockerfile
 
 .PHONY: build-forwarder
```

This would produce tags like `8.0.5-a1a6394cc5ae-357a339`. I don't know if we care that this is a) a bit inscrutable; b) not a proper semver (I think?).

Anyway, that's the general idea. Let's chat me-tonight/you-in-the-morning.

**Comment by gsleeman** [comment]
> I suggest is something like appending the current hash of this repo, like we do for operator versions. 

I've made a separate PR for this

File: audit-filter/README.md (hunk 5 above)
Comment (2uasimojo) [reply audit-filter/README.md:1]: Still need this.

**Comment by gsleeman** [comment]
Moved to another repo

**Pull request closed without merging**